backend/src/main/scala/bloop/io/ClasspathHasher.scala

```diff
@@ -1,284 +1,231 @@
+        hash.hash() match {
+          case BloopStamps.cancelledHash =>
+            if (cancelCompilation.isCompleted)
+              Task.now(FileHash.of(entry, BloopStamps.cancelledHash))
+            else {
+              // If the process that acquired it cancels the computation, try acquiring it again
+              logger
+                .warn(s"Unexpected hash computation of $entry was cancelled, restarting...")
+              acquireHashingEntry(entry).task
             }
-          )
-          found
+          case _ =>
+            Task.now(hash)
         }
+      }
 
-        val attrs = Files.readAttributes(jar.toPath, classOf[BasicFileAttributes])
-        val currentMetadata = (FileTime.fromMillis(IO.getModifiedTimeOrZero(jar)), attrs.size())
+    val onCancel = Task(isCancelled.compareAndSet(false, true)).unit
 
-        Option(definedMacrosJarCache.get(jar)) match {
-          case Some((metadata, hit)) if metadata == currentMetadata => jar -> hit
-          case _ =>
-            val detected = detectMacro(jar)
-            definedMacrosJarCache.put(jar, (currentMetadata, detected))
-            jar -> detected
+    val (computing, fromAnotherTask) =
+      classpath.toVector.map(path => acquireHashingEntry(path.underlying)).partition {
+        case _: Computed => true
+        case _: FromPromise => false
+      }
+
+    // first compute all the entries that are computed by this run
+    // then start waiting for entries that are computed by another tasks
+    val hashes = for {
+      computed <- Task.parSequenceN(parallelUnits)(computing.map(_.task).toList)
+      fromPromises <- Task.parSequenceN(parallelUnits)(fromAnotherTask.map(_.task).toList)
+    } yield computed ++ fromPromises
+
+    hashes
+      .map { result =>
+        if (isCancelled.get || cancelCompilation.isCompleted) {
+          cancelCompilation.trySuccess(())
+          Left(())
+        } else {
+          Right(result.toVector)
         }
       }
-    }
+      .doOnCancel(onCancel)
+  }
+}
 
-    Task.gatherUnordered(classpath.map(readJar(_)))
+object ClasspathHasher {
+  final val global: ClasspathHasher = new ClasspathHasher
+
+  private final case class JarMetadata(
+      modifiedTime: FileTime,
+      size: Long
+  )
+
+  /**
+   * File hash can be computed in 2 ways:
+   * 1) `Computed` - The file content is going to be hashed
+   * 2) `FromPromise` - The hash result is going to be obtained from the promise (another running
+   *     session of hashing is computing it)
+   *
+   * It's important to distinguish between them, because `Computed` case should be treated will
+   * higher priority as it needs CPU and `FromPromise` is just waiting on the result.
+   */
+  private sealed trait FileHashingResult {
+    def task: Task[FileHash]
   }
+  private final case class Computed(task: Task[FileHash]) extends FileHashingResult
+  private final case class FromPromise(task: Task[FileHash]) extends FileHashingResult
```

Comment:
Is scaladoc good enough to explain why this is needed?

Reply:
In the description you wrote:
```
remove previous mechanism which tried to share work between concurrent requests - it was causing too much mental overhead and was error prone. 
```
should we not have only computed since we don't want to depend on another task?